lib/run/frame.ts

```diff
@@ -17,83 +18,182 @@ export interface FrameOptions<T> {
 }
 
 export function createFrame<T>(options: FrameOptions<T>): Frame<T> {
-  let { operation, parent } = options;
-  let children = new Set<Frame>();
-  let block = createBlock(operation);
-  let context = Object.create(parent?.context ?? {});
-  let results = createEventStream<void, Result<void>>();
+  return evaluate<Frame<T>>(function* () {
+    let { operation, parent } = options;
+    let children = new Set<Frame>();
+    let context = Object.create(parent?.context ?? {});
+    let thunks: IteratorResult<Thunk, Result<T>>[] = [{
+      done: false,
+      value: $next(void 0),
+    }];
+
+    let crash: Error | undefined = void 0;
+
+    let interrupt = () => {};
+
+    let abort = (reason?: Error) => {
+      if (!frame.aborted) {
+        frame.aborted = true;
+        crash = reason;
+        thunks.unshift({ done: false, value: $abort() });
+        interrupt();
+      }
+    };
+
+    let [setResults, results] = yield* createValue<FrameResult<T>>();
+
+    let frame = yield* shiftSync<Frame<T>>((k) => {
+      let self: Frame<T> = create<Frame<T>>("Frame", { id: ids++, context }, {
+        createChild<X>(operation: () => Operation<X>) {
+          let child = createFrame<X>({ operation, parent: frame });
+          children.add(child);
+          evaluate(function* () {
+            yield* child;
+            children.delete(child);
+          });
+          return child;
+        },
+        enter: () => {
+          let task = createTask(self);
+          self.enter = () => task;
+          k.tail(self);
+          return task;
+        },
+        crash(error: Error) {
+          abort(error);
+          return frame;
+        },
+        destroy() {
+          abort();
+          return frame;
+        },
```

Comment:
Why do we need to return the `frame` here when we *must* already have it in scope when calling these methods?

Reply:
We're using the `Frame` as `Computation<FrameResult>` in this context, because the method signature is `destroy(): Computation<FrameResult>`

Reply:
I guess my question is: why?  Couldn't we:

```ts
frame.destroy();
yield* frame;
```

Reply:
I think it has to do with the history of `destroy()` being a computation in its own right.

`frame.destroy()` used to not do anything until you evaluated it. However, I made it a synchronous method to try and eliminate an intermediate generator.

https://github.com/thefrontside/effection/blob/e72b9ac1b38ff301f8afc044449007296b6c785f/lib/run/frame.ts#L69-L79

In a more general sense, I like methods that tell you when they are done rather than "fire and forget" followed by join at certain point. 

```ts
yield* frame.destroy();
```

When destruction is complete is ultimately the business of the frame, not the caller, so having it signal that destruction is fully settled when the computation returns feels like a good api.

Reply:
Makes sense :shipit: